automerge-backend/src/object_store.rs

```diff
@@ -52,6 +60,39 @@ impl ListState {
         }
     }
 
+    fn generate_diffs(&self) ->  Vec<Diff> {
+        let mut after = 0;
+        let ops_in_order = list_ops_in_order(&self.operations_by_elemid, &self.following).ok().unwrap_or(vec![]);
+        let mut diffs = vec![Diff {
+            action: DiffAction::CreateList(self.object_id.clone(), self.sequence_type.clone()),
+            conflicts: Vec::new(),
+        }];
+        diffs.extend(ops_in_order.iter().rev().filter_map(|(_,ops)|  {
+            ops.active_op()
+                .map(|op| {
+                  let tmp = list_op_to_assign_diff(&op.operation, &self.sequence_type, after)
+                      .map(|action|
+                          Diff {
+                            action: action,
+                            conflicts: ops.conflicts(),
+                          }
+                      );
+                  after += 1;
```

Comment:
This feels a bit gnarly, `map` generally being assumed to be side effect free. I _think_ it would be a good candidate for `Iter::scan`.